packages/controls/src/controls/group/group.ts

```diff
@@ -0,0 +1,310 @@
+import { z } from 'zod'
+
+import { mapValues } from '../../lib/functional'
+import { safeParse, type ParseResult } from '../../lib/zod'
+
+import { ControlDataTypeKey, type Data } from '../../common'
+import {
+  type CopyContext,
+  type MergeTranslatableDataContext,
+} from '../../context'
+import { type IntrospectionTarget } from '../../introspection'
+import {
+  SerializationSchema,
+  type DeserializedRecord,
+  type SerializedRecord,
+} from '../../serialization'
+
+import {
+  type DataType as DataType_,
+  type ResolvedValueType as ResolvedValueType_,
+  type ValueType as ValueType_,
+} from '../associated-types'
+import { ControlDefinition, serialize, type SchemaType } from '../definition'
+import { type SendMessage } from '../instance'
+import { ShapeDefinition } from '../shape/v1'
+import { ShapeV2Definition } from '../shape/v2'
+
+import { GroupControl } from './group-control'
+
+type ItemDefinition = ControlDefinition<string, unknown, any, any, any>
+type KeyDefinitions = Record<string, ItemDefinition>
+
+export type Config<Defs extends KeyDefinitions = KeyDefinitions> = {
+  readonly label?: string
+  readonly preferredLayout?:
+    | typeof Definition.Layout.Inline
+    | typeof Definition.Layout.Popover
+  readonly props: Defs
+}
+
+type ShapeData<C extends Config> = DataType_<
+  ShapeDefinition<{ type: C['props'] }>
+>
+
+type PropsData<C extends Config> = {
+  [K in keyof C['props']]?: DataType_<C['props'][K]>
+}
+
+type VersionedData<C extends Config> = {
+  [ControlDataTypeKey]:
+    | typeof Definition.v1DataType
+    | typeof ShapeV2Definition.v1DataType
+  value: PropsData<C>
+}
+
+type DataType<C extends Config> = ShapeData<C> | VersionedData<C>
```

Comment:
`Group` data is either `Shape` (v1) data or versioned data tagged with either `Group` or `ShapeV2` data type.